src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java

```diff
@@ -19,1749 +19,1718 @@
-			break;	
-			
-		case DIAG:
-			checkNumParameters(1);
-			checkMatrixParam(getFirstExpr());
-			output.setDataType(DataType.MATRIX);
-			if( id.getDim2() != -1 ) { //type known
-				if ( id.getDim2() == 1 ) 
-				{
-					//diag V2M
-					output.setDimensions(id.getDim1(), id.getDim1());
-				} 
-				else 
-				{
-					if (id.getDim1() != id.getDim2()) {
-						raiseValidateError("diag can either: (1) create diagonal matrix from (n x 1) matrix, or (2) take diagonal from a square matrix. "
-								+ "Error invoking diag on matrix with dimensions ("
-								+ id.getDim1() + "," + id.getDim2()
-								+ ") in " + this.toString(), conditional, LanguageErrorCodes.INVALID_PARAMETERS);
-					}
-					//diag M2V
-					output.setDimensions(id.getDim1(), 1);
-				}
-			}
-			output.setBlockDimensions (id.getRowsInBlock(), id.getColumnsInBlock());
-			output.setValueType(id.getValueType());
-			break;
-		case NROW:
-		case NCOL:
-		case LENGTH:
-			checkNumParameters(1);
-			checkMatrixFrameParam(getFirstExpr());
-			output.setDataType(DataType.SCALAR);
-			output.setDimensions(0, 0);
-			output.setBlockDimensions (0, 0);
-			output.setValueType(ValueType.INT);
-			break;
-
-		// Contingency tables
-		case TABLE:
-			
+        } catch (LanguageException e) {
+            throw new RuntimeException(e);
+        }
+        return paramExpression;
+    }
+
+    /**
+     * Validate parse tree : Process BuiltinFunction Expression in an assignment
+     * statement
+     *
+     * @throws LanguageException if LanguageException occurs
+     */
+    @Override
+    public void validateExpression(HashMap<String, DataIdentifier> ids, HashMap<String, ConstIdentifier> constVars, boolean conditional)
+            throws LanguageException {
+
+        for (int i = 0; i < _args.length; i++) {
+
+            if (_args[i] instanceof FunctionCallIdentifier) {
+                raiseValidateError("UDF function call not supported as parameter to built-in function call", false);
+            }
+
+            _args[i].validateExpression(ids, constVars, conditional);
+        }
+
+        // checkIdentifierParams();
+        String outputName = getTempName();
+        DataIdentifier output = new DataIdentifier(outputName);
+        output.setParseInfo(this);
+
+        if (this.getFirstExpr() == null) {
+            raiseValidateError("Function " + this + " has no arguments.", false);
+        }
+        Identifier id = this.getFirstExpr().getOutput();
+        output.setProperties(this.getFirstExpr().getOutput());
+        output.setNnz(-1); //conservatively, cannot use input nnz!
+        this.setOutput(output);
+
+        switch (this.getOpCode()) {
+            case EVAL:
+                checkNumParameters(Integer.MAX_VALUE);
+                output.setDataType(DataType.MATRIX);
+                output.setValueType(ValueType.DOUBLE);
+                output.setDimensions(1, 1);
+                output.setBlockDimensions(1000, 1000);
```

Comment:
Please derive the blocksize from the input if there is at least one matrix input or use the default block size from ConfigurationManager.getBlocksize()